fix(sentry): do not overwrite opts.Dsn

burmudar opened the pull request:
Upon `build()`, `opts.Dsn` is overwritten with `s.DSN` which is empty. Which means any explicitly Dsn set via opts is ignored.

I might be missing some original context here but I can't see a use case for DSN in `SentrySink` since it is defined in `sentry.ClientOptions` already

### Test plan
It's pretty difficult test since the options get copied during `build` and modified and passed onto the client. Otherwise, the unit tests still pass

jhchabran commented:
> I might be missing some original context here but I can't see a use case for DSN in SentrySink since it is defined in sentry.ClientOptions already

It's here! https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/conf/log_sinks.go

bobheadxi reviewed (changes requested):
I actually feel like the right path forward is to _remove_ `NewSentrySinkWithOptions`, and only export options through the `SentrySink` struct. Having `NewSentrySinkWithOptions` invites this exact confusion - the correct way to set a custom DSN is _not_ Sourcegraph-specific, it's the way all sinks are supposed to be configured in general:

```go
	liblog := log.Init(log.Resource{}, sglog.NewSentrySink())
	defer liblog.Sync()
	liblog.Update(func() log.SinksConfig {
		return log..SinksConfig{
			Sentry: &log.SentrySink{
				DSN: "my-dsn",
			},
		}
	})
```

burmudar commented:
Right, after  a bit of back and forth with myself I get it now 😆 

Then in line with this suggestion

> I actually feel like the right path forward is to _remove_ `NewSentrySinkWithOptions`, and only export options through the `SentrySink` struct. Having `NewSentrySinkWithOptions` invites this exact confusion - the correct way to set a custom DSN is _not_ Sourcegraph-specific, it's the way all sinks are supposed to be configured in general:
> 
> ```go
> 	liblog := log.Init(log.Resource{}, sglog.NewSentrySink())
> 	defer liblog.Sync()
> 	
> 	// Set initial configuration
> 	liblog.Update(func() log.SinksConfig {
> 		return log..SinksConfig{
> 			Sentry: &log.SentrySink{
> 				DSN: "my-dsn",
> 			},
> 		}
> 	})
> ```

Would be to add `SampleRate` as well in `SentrySink`.

A small point for exposing options - it allowed me to debug the sentry sink by passing in `sentrylib.ClientOptions{Debug: true}`

burmudar commented:
I've incorporated the 

> Right, after a bit of back and forth with myself I get it now 😆
> 
> Then in line with this suggestion
> 
> > I actually feel like the right path forward is to _remove_ `NewSentrySinkWithOptions`, and only export options through the `SentrySink` struct. Having `NewSentrySinkWithOptions` invites this exact confusion - the correct way to set a custom DSN is _not_ Sourcegraph-specific, it's the way all sinks are supposed to be configured in general:
> > ```go
> > 	liblog := log.Init(log.Resource{}, sglog.NewSentrySink())
> > 	defer liblog.Sync()
> > 	
> > 	// Set initial configuration
> > 	liblog.Update(func() log.SinksConfig {
> > 		return log..SinksConfig{
> > 			Sentry: &log.SentrySink{
> > 				DSN: "my-dsn",
> > 			},
> > 		}
> > 	})
> > ```
> 
> Would be to add `SampleRate` as well in `SentrySink`.
> 
> A small point for exposing options - it allowed me to debug the sentry sink by passing in `sentrylib.ClientOptions{Debug: true}`

I've added these values to SentrySink. Let me know what you folks think 😄

jhchabran commented on sinks_sentry.go:
```diff
@@ -24,48 +30,51 @@ type sentrySink struct {
 	}
 
-	if s.DSN == updatedDSN {
+	if s.DSN == updated.Sentry.DSN && s.SampleRate == updated.Sentry.SampleRate && s.Debug == updated.Sentry.Debug {
```
nit: can we find a better way to handle this? This is going to make us think it's working whereas it's not if we use a sentry option that isn't included here

burmudar replied:
Catching up here a bit. Why do we specifically export a few options and not just all the options ? Why can't we use something like:
```
type SentrySink struct {
   sentry.ClientOptions
}
```

and for the above comparison
```
if cmp.Equals(s.ClientOptions, updated.Sentry.ClientOptions) {
   return nil
}
```

jhchabran replied:
You're right, that'd be much simpler!

jhchabran commented on sinks_sentry.go:
```diff
@@ -24,48 +30,51 @@ type sentrySink struct {
+// To provide different values see `NewSentrySinkWith`
 func NewSentrySink() Sink {
-	return &sentrySink{SentrySink: SentrySink{options: sentrycore.DefaultSentryClientOptions}}
+	return &sentrySink{SentrySink: SentrySink{SampleRate: 0.1}}
```
Can we use instead use `sentrycore.DefaultSentryClientOptions` ?

jhchabran reviewed (changes requested):
The `default_sentry_client_opts.go` file is now entirely unused. Could we fix how we deal with defaults and make the code able to deal with options that we didn't account for?

jhchabran reviewed (approved):
Let's go!

bobheadxi commented on sinks_sentry.go:
```diff
@@ -12,10 +13,8 @@ import (
-
-	options sentry.ClientOptions
+	// ClientOptions expose various options to configure the Sentry client
+	sentry.ClientOptions
```
@jhchabran is this safe to make public? Why was it private to begin with? (just in case e.g. this struct is not read-only)

jhchabran replied:
Yes it is safe, it was unexported because we were passing the options through the constructor, which was enough. The current approach that @burmudar took gives more flexibility and enables to simply remove the additional constructor.

bobheadxi commented:
Ah... note that https://github.com/sourcegraph/log/pull/24 is not yet in downstream, which means that to introduce this change we will need to fix all the improper instantiations described in that PR. Feel free to revert https://github.com/sourcegraph/log/pull/24 to expedite things
